csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -301,27 +133,259 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
+
+  for (auto transform : transforms) {
+    if (transform->isA<Split>()) {
+      id = transform->as<Split>()->outer();
+    } else if (transform->isA<Merge>()) {
+      NVF_ERROR(
+          id == transform->as<Merge>()->outer(),
+          "Expected the sharding to be on the outer reshaped id.");
+      id = transform->as<Merge>()->out();
+    }
+  }
+  return id;
+}
+
+void propagateDIDTransform(
+    const TensorView* ref,
+    TensorView* tv,
+    const std::unordered_set<ParallelType>& selected_parallel_types,
+    PropagateDirection direction) {
+  tv->setDeviceMesh(ref->getDeviceMesh());
+
+  std::unordered_set<IterDomain*> device_ids;
+  std::copy_if(
+      ref->getLoopDomain().begin(),
+      ref->getLoopDomain().end(),
+      std::inserter(device_ids, device_ids.end()),
+      [&selected_parallel_types](IterDomain* id) {
+        return id->isDeviceDim() &&
+            selected_parallel_types.count(id->getParallelType());
+      });
+
+  if (device_ids.empty()) {
+    return;
+  }
+
+  std::unordered_set<IterDomain*> ignored_device_ids;
+  std::unordered_map<IterDomain*, IterDomain*> ref2target;
+  std::unordered_set<IterDomain*> input_ids;
+
+  // Finds all device IDs that are dependent on the given id
+  // and adds them to the ignored_device_ids set.
+  auto add_ignored_device_ids = [&ignored_device_ids,
+                                 &device_ids](IterDomain* id) {
+    for (auto device_id : device_ids) {
+      if (DependencyCheck::isDependencyOf(id, device_id)) {
+        ignored_device_ids.insert(device_id);
+      }
+    }
+  };
+
+  auto is_in_domain = [](IterDomain* id,
+                         const std::vector<IterDomain*>& domain) -> bool {
+    return std::find(domain.begin(), domain.end(), id) != domain.end();
+  };
+
+  if (direction == PropagateDirection::kForward) {
+    input_ids = getInputsInTargetDomain(
+        {device_ids.begin(), device_ids.end()},
+        {ref->getLogicalDomain().begin(), ref->getLogicalDomain().end()});
+    ref2target =
+        PairwiseLogicalDomainMap(ref, tv).mapProducerToConsumer(&input_ids);
+  }
+
+  if (direction == PropagateDirection::kBackward) {
+    input_ids = getInputsInTargetDomain(
+        {device_ids.begin(), device_ids.end()},
+        {ref->getMaybeRootDomain().begin(), ref->getMaybeRootDomain().end()});
+    ref2target =
+        PairwiseLogicalDomainMap(tv, ref).mapConsumerToProducer(&input_ids);
+  }
+
+  for (IterDomain* ref_id : ir_utils::filterByType<IterDomain>(input_ids)) {
+    if (ref2target.find(ref_id) == ref2target.end()) {
+      add_ignored_device_ids(ref_id);
+      continue;
+    }
+
+    IterDomain* target_id = ref2target.at(ref_id);
+
+    // For forward: target_id may not be logical.
+    // For backward: ref_id may not be logical.
+    IterDomain* outermost_ref_id = getOutermostLogicalId(ref_id, ref);
+    IterDomain* outermost_target_id = getOutermostLogicalId(target_id, tv);
```

Comment:
Not a request for this PR, but how do you like moving the follow-outermost logic into transformLoopTransform? 

While I understand your code now, I initially struggled to understand the meaning of the `ref2target` argument. It kind of captures what IterDomains are logically mapped but not exactly. In case of a spliting reshape, `h` in producer and `a` in consumer are added to `ref2target` but they don't even have the same extent. I agree that `h` in producer/consumer should map, and `d` in producer/consumer should map. However, `a` is an intermediate IterDomain on which we'll apply the split. It doesn't need to appear in ref2target, and IMO shouldn't.

Reply:
Additionally, I think a class like structure (something like a `DIDTransformPropagator`) may be cleaner instead of standalone methods.